nuxeo-features/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/action/RecomputeThumbnailsAction.java

```diff
@@ -0,0 +1,153 @@
+
+    public static final String PARAM_XPATH = "xpath";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        return Topology.builder()
+                       .addComputation(RecomputeThumbnailsComputation::new, //
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class RecomputeThumbnailsComputation extends AbstractBulkComputation {
+
+        protected String xpath;
+
+        public RecomputeThumbnailsComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        public void startBucket(String bucketKey) {
+            BulkCommand command = getCurrentCommand();
+            xpath = command.getParam(PARAM_XPATH);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+
+            DocumentModelList docs = loadDocuments(session, ids);
+
+            for (DocumentModel doc : docs) {
+                processDoc(session, doc);
+            }
+        }
+
+        protected void processDoc(CoreSession session, DocumentModel doc) {
+            Blob thumbnailBlob = getManagedThumbnail(doc);
+            if (thumbnailBlob == null) {
+                ThumbnailAdapter thumbnailAdapter = doc.getAdapter(ThumbnailAdapter.class);
+                if (thumbnailAdapter == null) {
+                    return;
+                }
+                thumbnailBlob = thumbnailAdapter.computeThumbnail(session);
+            }
+            if (thumbnailBlob != null) {
+                if (!doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                    doc.addFacet(ThumbnailConstants.THUMBNAIL_FACET);
+                }
+                doc.setPropertyValue(ThumbnailConstants.THUMBNAIL_PROPERTY_NAME, (Serializable) thumbnailBlob);
+            } else {
+                if (doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                    doc.setPropertyValue(ThumbnailConstants.THUMBNAIL_PROPERTY_NAME, null);
+                    doc.removeFacet(ThumbnailConstants.THUMBNAIL_FACET);
+                }
+            }
+            if (doc.isDirty()) {
+                doc.putContextData(VersioningService.VERSIONING_OPTION, VersioningOption.NONE);
+                doc.putContextData(VersioningService.DISABLE_AUTO_CHECKOUT, Boolean.TRUE);
+                doc.putContextData(DublinCoreListener.DISABLE_DUBLINCORE_LISTENER, Boolean.TRUE);
+                doc.putContextData(NotificationConstants.DISABLE_NOTIFICATION_SERVICE, Boolean.TRUE);
+                doc.putContextData("disableAuditLogger", Boolean.TRUE);
+                if (doc.isVersion()) {
+                    doc.putContextData(ALLOW_VERSION_WRITE, Boolean.TRUE);
+                }
+                doc.putContextData(THUMBNAIL_UPDATED, true);
+                session.saveDocument(doc);
```

Comment:
i am wondering about `doc.putContextData(THUMBNAIL_UPDATED, true);` in the case where `thumbnailBlob ` is 'null' and it hasn't any facet (doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) we will mark the document with THUMBNAIL_UPDATED even there  isn't any update (Audit purpose) wdyt and let us know the point of view of @bdelbosc / @ataillefer

Reply:
Yes, this is to avoid re-rentering the `CheckBlobUpdateListener` listener that will be triggered by  `session.saveDocument(doc);`, see https://github.com/nuxeo/nuxeo/blob/master/nuxeo-features/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/listener/CheckBlobUpdateListener.java#L59

`thumbnailBlob` being `null` whereas the thumbnail wasn't null previously will make the blob dirty (by setting the `ThumbnailConstants.THUMBNAIL_PROPERTY_NAME` property to `null`), thus trigger the save, so we need it, as it is done in the code from which this was taken: https://github.com/nuxeo/nuxeo/blob/master/nuxeo-features/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/listener/UpdateThumbnailListener.java#L60.